src/plonkish/backend/hyperplonk.rs

```diff
@@ -0,0 +1,441 @@
+            num_by_phase[phase as usize] += 1;
+            num_by_phase
+        },
+    )
+}
+
+// This function maps each element in the phases slice to its index within its specific phase.
+// For example, if the phases slice is [0, 1, 0, 1, 2, 2], then the output vector will be
+// [0, 0, 1, 1, 0, 1].
+fn idx_in_phase(phases: &Vec<usize>) -> Vec<usize> {
+    phases
+        .iter()
+        .copied()
+        .scan(vec![0; num_phases(phases)], |state, phase| {
+            let index = state[phase as usize];
+            state[phase as usize] += 1;
+            Some(index)
+        })
+        .collect()
+}
+
+// This function maps each element in the phases slice to its index within the circuit, given an
+// offset For example, if the phases slice is [0, 1, 0, 1, 2, 2], and the offset is 3, then the
+// output vector will be [3, 5, 4, 6, 7, 8], i.e. [3+0+0, 3+2+0, 3+0+1, 3+2+1, 3+4+0, 3+4+1], i.e.
+// [offset+phase_offset+index]
+fn idx_order_by_phase(phases: &Vec<usize>, offset: usize) -> Vec<usize> {
+    phases
+        .iter()
+        .copied()
+        .scan(phase_offsets(phases), |state, phase| {
+            let index = state[phase as usize];
+            state[phase as usize] += 1;
+            Some(offset + index)
+        })
+        .collect()
+}
+
+// get vector of non-selector advice column phases
+fn non_selector_advice_phases<F: Field>(circuit: &Circuit<F>) -> Vec<usize> {
+    circuit
+        .columns
+        .iter()
+        .filter(|column| {
+            column.ctype == ColumnType::Advice && !column.annotation.starts_with("step selector")
+        })
+        .map(|column| column.phase)
+        .collect::<Vec<usize>>()
+}
+
+// This function computes the offsets for each phase.
+// For example, if the phases slice is [0, 1, 0, 1, 2, 2], then the output vector will be
+// [0, 2, 4].
+fn phase_offsets(phases: &Vec<usize>) -> Vec<usize> {
+    num_by_phase(phases)
+        .into_iter()
+        .scan(0, |state, num| {
+            let offset = *state;
+            *state += num;
+            Some(offset)
+        })
+        .collect()
+}
+
+pub struct ChiquitoCircuit<F, TraceArgs> {
+    // TraceArgs is for trace_gen, which is not used here; we only care about placement
+    k: usize,
+    instances: Vec<Vec<F>>, /* outter vec has length 1, inner vec has length equal to number of
+                             * exposed signals */
+    circuit: Circuit<F>,
+    num_witness_polys: Vec<usize>,
+    advice_idx_in_phase: Vec<usize>,
+    assignment_generator: AssignmentGenerator<F, TraceArgs>,
+    trace_witness: TraceWitness<F>,
+    all_uuids: Vec<UUID>,      // the same order as self.circuit.columns
+    fixed_uuids: Vec<UUID>,    // the same order as self.circuit.columns
+    selector_uuids: Vec<UUID>, // the same order as self.circuit.columns
+    non_selector_advice_uuids: Vec<UUID>, // the same order as self.circuit.columns
+    non_selector_advice_uuids_by_phase: HashMap<usize, Vec<UUID>>,
```

Comment:
Remove the selector. Chiquito selector and non selector will both be advice. They are not fixed.

Replace with a HashMap as it's O(logN) while vector fine is O(N)

Reply:
Note that we need to find the index of the UUID within the collection and keep the order, but HashMap doesn't keep any order, so we still have to use a Vec